File: compare-sort-algorithms/src/QuickSort.java
import java.io.*;
import java.util.*;

public class QuickSort {
    
    public static void sort(int[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);
            
            // Recursively sort elements before and after partition
            sort(arr, low, pivotIndex - 1);
            sort(arr, pivotIndex + 1, high);
        }
    }
    
    public static void sort(int[] arr) {
        sort(arr, 0, arr.length - 1);
    }
    
    private static int partition(int[] arr, int low, int high) {
        // Choose the rightmost element as pivot
        int pivot = arr[high];
        
        // Index of smaller element
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (arr[j] <= pivot) {
                i++;
                // Swap elements
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        
        // Place pivot in correct position
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        
        return i + 1;
    }
    
    public static int[] readNumbersFromFile(String filename) throws IOException {
        List<Integer> numbers = new ArrayList<>();
        
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                try {
                    numbers.add(Integer.parseInt(line.trim()));
                } catch (NumberFormatException e) {
                    // Skip non-integer lines
                    continue;
                }
            }
        } catch (FileNotFoundException e) {
            System.err.println("Error: " + filename + " not found. Please run generate_data.py first.");
            System.exit(1);
        }
        
        return numbers.stream().mapToInt(Integer::intValue).toArray();
    }
    
    public static boolean isSorted(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            if (arr[i] > arr[i + 1]) {
                return false;
            }
        }
        return true;
    }
    
    public static void main(String[] args) {
        System.out.println("Java Quick Sort Performance Test");
        System.out.println("================================");
        
        try {
            // Get input and output filenames from command line arguments
            String inputFile = "datasets/random_list.txt";  // Fixed path to include datasets directory
            String outputFile = "results/results_java.txt"; // Fixed path to include results directory
            
            if (args.length > 0) {
                inputFile = args[0];
            }
            
            if (args.length > 1) {
                outputFile = args[1];
            }
            
            System.out.println("Using dataset: " + inputFile);
                
            // Read data from file
            System.out.println("Reading data from file...");
            int[] data = readNumbersFromFile(inputFile);
            System.out.println("Data size: " + data.length + " integers");
            
            // Create a copy for sorting (to preserve original)
            int[] dataCopy = Arrays.copyOf(data, data.length);
            
            // Measure sorting time
            System.out.println("Starting Quick Sort...");
            long startTime = System.nanoTime();
            sort(dataCopy);
            long endTime = System.nanoTime();
            
            double executionTime = (endTime - startTime) / 1_000_000_000.0; // Convert to seconds
            
            // Verify the array is sorted
            boolean sorted = isSorted(dataCopy);
            
            // Results
            System.out.println("Sorting completed: " + (sorted ? "SUCCESS" : "FAILED"));
            System.out.printf("Execution time: %.6f seconds%n", executionTime);
            System.out.printf("Elements per second: %.0f%n", data.length / executionTime);
            
            // Write results to the output file
            try (PrintWriter writer = new PrintWriter(new FileWriter(outputFile))) {
                writer.println("Java Quick Sort Results");
                writer.println("Data size: " + data.length);
                writer.printf("Execution time: %.6f seconds%n", executionTime);
                writer.printf("Elements per second: %.0f%n", data.length / executionTime);
                writer.println("Sorted correctly: " + sorted);
            }
            
            System.out.println("Results saved to " + outputFile);
            
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: compare-sort-algorithms/src/quick_sort.js
#!/usr/bin/env node
/**
 * Quick Sort implementation in JavaScript with performance measurement.
 */
const fs = require('fs');

class QuickSort {
    static sort(arr, low = 0, high = arr.length - 1) {
        if (low < high) {
            const pivotIndex = this.partition(arr, low, high);
            
            // Recursively sort elements before and after partition
            this.sort(arr, low, pivotIndex - 1);
            this.sort(arr, pivotIndex + 1, high);
        }
    }
    
    static partition(arr, low, high) {
        // Choose the rightmost element as pivot
        const pivot = arr[high];
        
        // Index of smaller element
        let i = low - 1;
        
        for (let j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (arr[j] <= pivot) {
                i++;
                // Swap elements
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
        
        // Place pivot in correct position
        [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
        return i + 1;
    }
}

function readNumbersFromFile(filename = 'random_list.txt') {
    try {
        const data = fs.readFileSync(filename, 'utf8');
        return data.trim().split('\n')
            .map(line => parseInt(line.trim()))
            .filter(num => !isNaN(num));
    } catch (error) {
        if (error.code === 'ENOENT') {
            console.error(`Error: ${filename} not found. Please run generate_data.py first.`);
            process.exit(1);
        }
        throw error;
    }
}

function isSorted(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

function main() {
    console.log('JavaScript Quick Sort Performance Test');
    console.log('======================================');
    
    // Get input and output filenames from command line arguments
    const inputFile = process.argv[2] || 'random_list.txt';
    const outputFile = process.argv[3] || 'results_javascript.txt';
    
    console.log(`Using dataset: ${inputFile}`);
    
    // Read data from file
    console.log('Reading data from file...');
    const data = readNumbersFromFile(inputFile);
    console.log(`Data size: ${data.length} integers`);
    
    // Create a copy for sorting (to preserve original)
    const dataCopy = [...data];
    
    // Measure sorting time
    console.log('Starting Quick Sort...');
    const startTime = process.hrtime.bigint();
    QuickSort.sort(dataCopy);
    const endTime = process.hrtime.bigint();
    
    const executionTime = Number(endTime - startTime) / 1_000_000_000; // Convert to seconds
    
    // Verify the array is sorted
    const sorted = isSorted(dataCopy);
    
    // Results
    console.log(`Sorting completed: ${sorted ? 'SUCCESS' : 'FAILED'}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
    console.log(`Elements per second: ${Math.round(data.length / executionTime)}`);
    
    // Save results to the specified output file
    const results = `JavaScript Quick Sort Results
Data size: ${data.length}
Execution time: ${executionTime.toFixed(6)} seconds
Elements per second: ${Math.round(data.length / executionTime)}
Sorted correctly: ${sorted}
`;
    
    fs.writeFileSync(outputFile, results);
    console.log(`Results saved to ${outputFile}`);
}

if (require.main === module) {
    main();
}
